### PR TITLE
fix shipment dependency

### PR DESCRIPTION
**Pull request by DimanKadykov**
- Developer(s): @pyvovarov-s

- Ticket: https://spryker.atlassian.net/browse/ECO-3339

- Academy PR: ACADEMY_URL_HERE

- Release Group: https://release.spryker.com/release-groups/view/2268

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Braintree             | minor                 |                       |

#### Release Notes

{skip}

-----------------------------------------

##### Change log

Improvements

Adjusted this module to allow new `spryker/shipment: "^8.0.0"`, `spryker/oms: "^11.0.0"` and `spryker/payment: "^5.0.0"` versions.

### CODE REVIEW

File: src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php [hunk 1]
```diff
@@ -105,14 +106,16 @@ protected function createAvailableShipmentChoiceList(QuoteTransfer $quoteTransfe
         $shipmentMethods = [];
 
         $shipmentMethodsTransfer = $this->getAvailableShipmentMethods($quoteTransfer);
-        foreach ($shipmentMethodsTransfer->getMethods() as $shipmentMethodTransfer) {
-            if (!isset($shipmentMethods[$shipmentMethodTransfer->getCarrierName()])) {
-                $shipmentMethods[$shipmentMethodTransfer->getCarrierName()] = [];
+        foreach ($shipmentMethodsTransfer->getShipmentMethods() as $shipmentMethodsTransfer) {
+            foreach ($shipmentMethodsTransfer->getMethods() as $shipmentMethodTransfer) {
+                if (!isset($shipmentMethods[$shipmentMethodTransfer->getCarrierName()])) {
```
Comment (dereuromark) [inline src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: You are using `$shipmentMethodTransfer->getCarrierName()` 3x
You could store this in local var and use that 3x then.

Comment (DimanKadykov) [reply src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: > You are using `$shipmentMethodTransfer->getCarrierName()` 3x
> You could store this in local var and use that 3x then.

that's not an improvement. https://boyet.com/blog/local-variables-are-evil-or-just-irritating/

Comment (dereuromark) [reply src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: It actually is an improvement, we do this.

Comment (dereuromark) [reply src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: That article was 10 years ago.
The code also becomes more readable and maintainable, which is usually a plus:
```php
//$shipmentMethods[$shipmentMethodTransfer->getCarrierName()][$description] = ...
$shipmentMethods[$carrierName][$description] = ...
```

Comment (DimanKadykov) [reply src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: OK i did that)

File: composer.json [hunk 2]
```diff
@@ -12,7 +12,7 @@
     "spryker/gui": "^3.0.0",
     "spryker/kernel": "^3.33.0",
     "spryker/money": "^2.0.0",
-    "spryker/payment": "^4.0.0",
+    "spryker/payment": "^5.0.0",
```
Comment (tamasnyulas) [inline composer.json]: This is a BC break, the original idea was to only allow the new major version.

File: src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientBridge.php [hunk 3]
```diff
@@ -28,10 +29,10 @@ public function __construct($shipmentClient)
...
      * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
      *
-     * @return \Generated\Shared\Transfer\ShipmentMethodsTransfer
+     * @return \Generated\Shared\Transfer\ShipmentMethodsCollectionTransfer
      */
-    public function getAvailableMethods(QuoteTransfer $quoteTransfer): ShipmentMethodsTransfer
+    public function getAvailableMethods(QuoteTransfer $quoteTransfer): ShipmentMethodsCollectionTransfer
     {
-        return $this->shipmentClient->getAvailableMethods($quoteTransfer);
+        return $this->shipmentClient->getAvailableMethodsByShipment($quoteTransfer);
```
Comment (tamasnyulas) [inline src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientBridge.php]: This is also a BC break as this method doesn't exist in the old versions.

File: src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php [hunk 4]
```diff
@@ -105,14 +106,18 @@ protected function createAvailableShipmentChoiceList(QuoteTransfer $quoteTransfe
         $shipmentMethods = [];
 
         $shipmentMethodsTransfer = $this->getAvailableShipmentMethods($quoteTransfer);
-        foreach ($shipmentMethodsTransfer->getMethods() as $shipmentMethodTransfer) {
-            if (!isset($shipmentMethods[$shipmentMethodTransfer->getCarrierName()])) {
-                $shipmentMethods[$shipmentMethodTransfer->getCarrierName()] = [];
+        foreach ($shipmentMethodsTransfer->getShipmentMethods() as $shipmentMethodsTransfer) {
```
Comment (tamasnyulas) [inline src/SprykerEco/Yves/Braintree/Form/DataProvider/CheckoutShipmentFormDataProvider.php]: I guess this part is also based on the new major version only.

File: composer.json (hunk 2 above)
Comment (dereuromark) [reply composer.json]: This triggers a (cascading) major here, yeah.

File: src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientBridge.php (hunk 3 above)
Comment (DimanKadykov) [reply src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientBridge.php]: yes. the task was about fixing dependency with latest shipment

**Comment by dereuromark** [comment]
Please also add a template here then, this is needed to be able to review.

**Comment by pyvovarov-s** [comment]
All issues fixed and Pr overview is ready as well as Release group

File: src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientInterface.php [hunk 5]
```diff
@@ -8,6 +8,7 @@
 namespace SprykerEco\Yves\Braintree\Dependency\Client;
 
 use Generated\Shared\Transfer\QuoteTransfer;
+use Generated\Shared\Transfer\ShipmentMethodsCollectionTransfer;
```
Comment (tamasnyulas) [inline src/SprykerEco/Yves/Braintree/Dependency/Client/BraintreeToShipmentClientInterface.php]: Seems like you have some leftovers

File: src/SprykerEco/Zed/Braintree/Business/Payment/Transaction/CaptureItemsTransaction.php [hunk 6]
```diff
@@ -207,7 +207,7 @@ protected function addOrderIdToAttributes(array $attributes): array
     protected function getShipmentExpenses(iterable $expenseTransfers): int
     {
         foreach ($expenseTransfers as $expenseTransfer) {
-            if ($expenseTransfer->getType() === ShipmentConstants::SHIPMENT_EXPENSE_TYPE) {
+            if ($expenseTransfer->getType() === ShipmentConfig::SHIPMENT_EXPENSE_TYPE) {
```
Comment (tamasnyulas) [inline src/SprykerEco/Zed/Braintree/Business/Payment/Transaction/CaptureItemsTransaction.php]: This is a BC break. See my suggestion for the fix:

```suggestion
            if ($expenseTransfer->getType() === static::SHIPMENT_EXPENSE_TYPE) {
```

... and on top of the class:
```
/**
 * @see \Spryker\Shared\Shipment\ShipmentConfig::SHIPMENT_EXPENSE_TYPE|\Spryker\Shared\Shipment\ShipmentConstants::SHIPMENT_EXPENSE_TYPE
 */
protected const SHIPMENT_EXPENSE_TYPE = 'SHIPMENT_EXPENSE_TYPE';
```

Please fix all occurrences in this PR.

**Pull request merged**